Add non-root temporary FHS chrootenv

abbradar opened the pull request:
This adds `buildFHSUserEnv`, a version of `buildFHSChrootEnv` which uses recent Linux namespaces support in kernel to achieve chrooted FHS environment without root. It uses temporary directory to build an environment (opposed to `ChrootEnv`, which builds a global environment for the system) and destroys it after chrooted program finishes running. Chrooting is done in a Ruby script with several FFI calls (such as `unshare`). This brings better experience for software that doesn't actually need persistent environment between runs, such as Steam -- you can now just run `steam` without any `-chrootenv` preparation. This also splits `buildFHSChrootEnv` into `buildFHSEnv`, which creates actual directory structure, and `chrootFHSEnv`, which makes run scripts for a given environment.

As usual, every naming I've used in this is awful ^_^

abbradar committed:
buildFHSChrootEnv: split environment and chroot scripts

abbradar committed:
userFHSEnv: add build tool

abbradar committed:
steam-chrootenv: use UserEnv

jagajaga commented:
:+1: :beeers:

jagajaga commented:
I think, we should mention somewhere that the behavior of `steam-chrootenv` has changed. Do you agree?

abbradar commented:
Good idea, I just don't know where -- nobody would read build log, for example (maybe it's the best we can do, though).
Also it seems like I've broken several paths in `chrootenv`, I'll fix them shortly.

jagajaga commented:
And also, maybe we should rename `steam` to smth like `steam main binary but not working one without chrootEnv`, and this `steamChrootEnv` to `steam`?

abbradar commented:
I've wanted to do this actually, but feared Backwards Compatibility beast. We can just do `steam = steamChrootEnv` though.

abbradar commented:
(I meant `steamChrootEnv = steam`)

jagajaga commented:
IMO let's add a some kind of alert, that steam is now renamed (changed). This is a situation, when backwards compatibility is bad.

abbradar committed:
steam: rename, add a warning

abbradar commented:
Fixed `ChrootEnv`, updated `chroot-user` script to relay signals to its child and to ensure cleanup, renamed `steamChrootEnv` to `steam` and added a warning.

jagajaga commented:
Seems to be nice now!
But about warning: when user download steamChrootEnv from cache - will he see this warning?

abbradar committed:
build-fhs-*: prefer local build

abbradar commented:
Nice catch, not now ~_^. `buildEnv` does the same thing, I don't think we should build these environments on Hydra, too.
P.S.: Just remembered, Steam is unfree, so these builds shouldn't be cached at all.

jagajaga commented:
So it's safe to merge now, I think.

abbradar commented:
Indeed, everything seems to be okay. Thanks for review and advices!

wmertens commented on pkgs/build-support/build-fhs-userenv/default.nix:
```diff
@@ -0,0 +1,37 @@
+    destination = "/bin/chroot-user";
+    text = ''
+      #! ${ruby}/bin/ruby
+      ${builtins.readFile ./chroot-user.rb}
```
Not entirely sure but I think this may do the read at evaluation time.  @edolstra?

abbradar replied:
Hm, I thought it would, too, but this file lies in `nixpkgs` and I supposed that we rely on it being unchanged while evaluating (otherwise lazy `import`s would be bad, too). Or have I completely misunderstood the issue?

wmertens replied:
Well I was thinking that it looked similar to
https://www.mail-archive.com/nix-dev@lists.science.uu.nl/msg14656.html
where cacert gets built at eval time but must say I don't fully understand
how that happens. Maybe the problem there was the reference to pkgs.cacert.

On Thu Feb 05 2015 at 9:57:30 PM Nikolay Amiantov <notifications@github.com>
wrote:

> In pkgs/build-support/build-fhs-userenv/default.nix
> <https://github.com/NixOS/nixpkgs/pull/6177#discussion_r24198705>:
>
> > @@ -0,0 +1,37 @@
> > +{ writeTextFile, stdenv, ruby } : { env, runScript } :
> > +
> > +let
> > +  name = env.pname;
> > +
> > +  # Sandboxing script
> > +  chroot-user = writeTextFile {
> > +    name = "chroot-user";
> > +    executable = true;
> > +    destination = "/bin/chroot-user";
> > +    text = ''
> > +      #! ${ruby}/bin/ruby
> > +      ${builtins.readFile ./chroot-user.rb}
>
> Hm, I thought it would, too, but this file lies in nixpkgs and I supposed
> that we rely on it being unchanged while evaluating (otherwise lazy imports
> would be bad, too). Or have I completely misunderstood the issue?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/6177/files#r24198705>.
>

abbradar replied:
If I understood correctly, the problem was because of user reading a big file in his configuration on evaluation and writing it to the output. Because our `writeTextFile` is implemented via `echo -n`, this resulted in `Argument list too long` error/ However, here I read a script from `nixpkgs`, so we (as maintainers) control the length of file, and if it gets too large after some change we'll invent another way to write it before it would break anything for users.

Another way of fixing this would be to change `writeTextFile` to output the file to the nix store first via `toFile` builtin, and then use `cat` in its build script -- then we would have no restriction on `text` size.

jagajaga commented:
Oh. Just realized, that there is no need in camelCase.
Maybe let's use `steam-original` instead?
IMO

abbradar commented:
Ah, it's really so -- I'm just accustomed to it after writing loads of Haskell ^_^ I'll rename it.

jagajaga commented:
I know that feel :)
